sprokit/src/bindings/python/sprokit_logging.py

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+"""
+A thin wrapper around python logging.
+"""
+from __future__ import print_function, unicode_literals, absolute_import
+import sys
+import six
+import os
+import logging
+
+
+def _configure_logging():
+    """
+    Configures python logging to using KWIVER / SPROKIT environment variables
+    """
+    # Use the C++ logging level by default, but allow python to be different
+    cxx_level = os.environ.get('KWIVER_DEFAULT_LOG_LEVEL', 'DEBUG')
+    py_level = os.environ.get('KWIVER_PYTHON_DEFAULT_LOG_LEVEL', cxx_level)
+    level = getattr(logging, py_level.upper())
+    logfmt = '%(name)s:%(levelname)s: %(message)s'
+
+    USE_COLOR = False
+    try:
+        import coloredlogs
+    except ImportError:
+        USE_COLOR = False
+
+    if USE_COLOR:
+        coloredlogs.DEFAULT_FIELD_STYLES['name']['color'] = 'blue'
+        coloredlogs.DEFAULT_LEVEL_STYLES['debug']['color'] = 'cyan'
+        # coloredlogs.DEFAULT_LEVEL_STYLES['warn']['color'] = 'yellow'
+        coloredlogs.install(level=level, fmt=logfmt)
+    else:
+        logging.basicConfig(format=logfmt, level=level)
+
+
+def print_exc(exc_info=None):
+    """
+    Prints a highly visible exception.
+
+    Example:
+        >>> try:
+        >>>     raise Exception('foobar')
+        >>> except Exception as ex:
+        >>>     import sys
+        >>>     exc_info = sys.exc_info()
+        >>>     print_exc(exc_info)
+    """
+    import traceback
+    if exc_info is None:
+        exc_info = sys.exc_info()
+    tbtext = ''.join(traceback.format_exception(*exc_info))
+
+    lines = [
+        '',
+        '┌───────────',
+        '│ EXCEPTION:',
+        '',
+        tbtext,
+        '└───────────',
+        ''
+    ]
+    text = '\n'.join(lines)
+    print(text)
+
+
+def exc_report(func):
+    """
+    Prints a message if an exception occurs in the decorated function.
+
+    Modules called from C++ (e.g. pybind11) are not gaurenteed to print
+    exceptions if they occur (unless the C++ program does the work). This
+    decorator can be used as a workaround to ensure that there is some
+    indication of when Python code raises an exception.
+    """
+    def _exc_report_wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception:
+            print_exc()
+            raise
+    return _exc_report_wrapper
+
+
+class SprokitLogger(object):
+    """
+    Lighweight wrapper around python's logging.Logger class.
+    """
+    @exc_report
+    def __init__(self, name):
+        self._logger = logging.getLogger(name)
+        self.log('debug', 'created logger for ' + name)
+
+    @exc_report
+    def log(self, level, msg):
+        if isinstance(level, six.string_types):
+            level = getattr(logging, level.upper())
+        self._logger.log(level, msg)
```

Comment:
Could this be routed through the vital logger?

Reply:
Hmm, it might be possible to write the appropriate C++ bindings. However, I think that may add unnecessary complexity. Also, Python's logging system is very mature and well suited to logging in a python environment (e.g. it can produce line numbers in the python source files or be configured to produce colored output). We may be able to implement a special Handler class that routes to the vital logger, but my initial feeling are that this is unnecessary.

Reply:
There is a use case we have encountered where our software is embedded in a larger system and the customer wants all log messages to go through that logger so they can manage and capture the log stream. This was done by adding a project specific logger back end, but it does require that all log messages be run through the main c++ logger. We do not need this support at this time, but it is worth making sure the architecture can support it.